{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import time\n",
    "import gzip\n",
    "import itertools\n",
    "import sys\n",
    "import json\n",
    "import pysam\n",
    "from Levenshtein import distance\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> Input Files </H1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "samp=\"NW072-WT1-210123_S1_L001\"\n",
    "RE=\"AAGCTT\"\n",
    "prbsALL = json.load(open('./PROBE_DEFS/NWProbeDef171116.py'))\n",
    "RefGenome = \"Genomes/HG19_BDGP5_PhiX_K12.fa\"\n",
    "useUMI=False\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################################################################################################################\n",
    "## LOAD REQUIRED DATASETS\n",
    "############################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDGP5:chr2L\n",
      "BDGP5:chr2LHet\n",
      "BDGP5:chr2R\n",
      "BDGP5:chr2RHet\n",
      "BDGP5:chr3L\n",
      "BDGP5:chr3LHet\n",
      "BDGP5:chr3R\n",
      "BDGP5:chr3RHet\n",
      "BDGP5:chr4\n",
      "BDGP5:chrU\n",
      "BDGP5:chrUextra\n",
      "BDGP5:chrX\n",
      "BDGP5:chrXHet\n",
      "BDGP5:chrYHet\n",
      "BDGP5:chrdmel_mitochondrion_genome\n",
      "K12:chrEcoli\n",
      "HG19:chr1\n",
      "HG19:chr10\n",
      "HG19:chr11\n",
      "HG19:chr12\n",
      "HG19:chr13\n",
      "HG19:chr14\n",
      "HG19:chr15\n",
      "HG19:chr16\n",
      "HG19:chr17\n",
      "HG19:chr18\n",
      "HG19:chr19\n",
      "HG19:chr2\n",
      "HG19:chr20\n",
      "HG19:chr21\n",
      "HG19:chr22\n",
      "HG19:chr3\n",
      "HG19:chr4\n",
      "HG19:chr5\n",
      "HG19:chr6\n",
      "HG19:chr7\n",
      "HG19:chr8\n",
      "HG19:chr9\n",
      "HG19:chrMT\n",
      "HG19:chrX\n",
      "HG19:chrY\n",
      "HG19:chrGL000207.1\n",
      "HG19:chrGL000226.1\n",
      "HG19:chrGL000229.1\n",
      "HG19:chrGL000231.1\n",
      "HG19:chrGL000210.1\n",
      "HG19:chrGL000239.1\n",
      "HG19:chrGL000235.1\n",
      "HG19:chrGL000201.1\n",
      "HG19:chrGL000247.1\n",
      "HG19:chrGL000245.1\n",
      "HG19:chrGL000197.1\n",
      "HG19:chrGL000203.1\n",
      "HG19:chrGL000246.1\n",
      "HG19:chrGL000249.1\n",
      "HG19:chrGL000196.1\n",
      "HG19:chrGL000248.1\n",
      "HG19:chrGL000244.1\n",
      "HG19:chrGL000238.1\n",
      "HG19:chrGL000202.1\n",
      "HG19:chrGL000234.1\n",
      "HG19:chrGL000232.1\n",
      "HG19:chrGL000206.1\n",
      "HG19:chrGL000240.1\n",
      "HG19:chrGL000236.1\n",
      "HG19:chrGL000241.1\n",
      "HG19:chrGL000243.1\n",
      "HG19:chrGL000242.1\n",
      "HG19:chrGL000230.1\n",
      "HG19:chrGL000237.1\n",
      "HG19:chrGL000233.1\n",
      "HG19:chrGL000204.1\n",
      "HG19:chrGL000198.1\n",
      "HG19:chrGL000208.1\n",
      "HG19:chrGL000191.1\n",
      "HG19:chrGL000227.1\n",
      "HG19:chrGL000228.1\n",
      "HG19:chrGL000214.1\n",
      "HG19:chrGL000221.1\n",
      "HG19:chrGL000209.1\n",
      "HG19:chrGL000218.1\n",
      "HG19:chrGL000220.1\n",
      "HG19:chrGL000213.1\n",
      "HG19:chrGL000211.1\n",
      "HG19:chrGL000199.1\n",
      "HG19:chrGL000217.1\n",
      "HG19:chrGL000216.1\n",
      "HG19:chrGL000215.1\n",
      "HG19:chrGL000205.1\n",
      "HG19:chrGL000219.1\n",
      "HG19:chrGL000224.1\n",
      "HG19:chrGL000223.1\n",
      "HG19:chrGL000195.1\n",
      "HG19:chrGL000212.1\n",
      "HG19:chrGL000222.1\n",
      "HG19:chrGL000200.1\n",
      "HG19:chrGL000193.1\n",
      "HG19:chrGL000194.1\n",
      "HG19:chrGL000225.1\n",
      "HG19:chrGL000192.1\n",
      "PhiX:chrPhix\n"
     ]
    }
   ],
   "source": [
    "refGenome={}\n",
    "for chromo in SeqIO.parse(open(RefGenome,\"r\"), \"fasta\"):\n",
    "    refGenome[chromo.id] = str(chromo.seq).upper()\n",
    "    print(chromo.id)\n",
    "\n",
    "\n",
    "lstBlackListed = {}\n",
    "for x in open(\"../DATA/ENCODE_Blacklist.regions.tsv\"):\n",
    "    [SRC,chrom,chromStart,chromEnd,name] = x[:-1].split(\"\\t\")\n",
    "    if SRC != 'SRC': \n",
    "       a = int((int(chromStart)/1000) - 1)\n",
    "       b = int((int(chromEnd)/1000) + 2 )\n",
    "        \n",
    "       for i in range(a,b): \n",
    "           lstBlackListed[chrom+\":\"+ str(i)] = True\n",
    "probes={}\n",
    "for xprb in prbsALL: \n",
    "    prb = prbsALL[xprb]\n",
    "    szPrb = prb[4] - prb[3]\n",
    "    if prb[2] == \"-\": \n",
    "       posRE = refGenome[prb[6]+\":\" +prb[1]].rfind(RE,0,prb[3])\n",
    "       xprbStart=prb[4]\n",
    "       xprbEnd=prb[3]\n",
    "       xfragEnd=posRE \n",
    "    else: \n",
    "       posRE = refGenome[prb[6]+\":\"+prb[1]].find(RE,prb[4])\n",
    "       xprbStart=prb[3]\n",
    "       xprbEnd=prb[4]\n",
    "       xfragEnd=posRE + 6\n",
    "    entry = [prb[0], xprbStart, xprbEnd, xfragEnd, prb[2]]\n",
    "    for i in range( int(prb[3]/1000) - 3, int(prb[3]/1000) + 4):   \n",
    "        binid=prb[1] + \":\" + str(i)    \n",
    "        if not binid in probes: \n",
    "           probes[binid] = []\n",
    "        probes[binid].append(entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################################################################################################################\n",
    "## LOAD Sam files\n",
    "############################################################################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Counting R2', 'NW072-WT1-210123_S1_L001', 0]\n",
      "['Counting R2', 'NW072-WT1-210123_S1_L001', 1000000.0]\n",
      "['Counting R2', 'NW072-WT1-210123_S1_L001', 2000000.0]\n",
      "['Counting R2', 'NW072-WT1-210123_S1_L001', 3000000.0]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "valid={}\n",
    "statsLib={}\n",
    "statsFrag={}\n",
    "statsPrbs={}\n",
    "umis={}\n",
    "\n",
    "#### Count UMI's\n",
    "iprog=0\n",
    "if useUMI: \n",
    "   for u in SeqIO.parse(gzip.open(\"../UPLOADS/NW_ST1/\"+samp+\"_I2.fastq.gz\", \"rt\"),\"fastq\"): \n",
    "       umis[u.id] = str(u.seq)\n",
    "       if iprog % 100000 == 0: \n",
    "          print ([\"Loading UMI\", samp, iprog])\n",
    "       iprog += 1\n",
    "\n",
    "#### Count input number of reads from FASTQ\n",
    "rd2ct=0\n",
    "iprog=0\n",
    "for x in gzip.open(\"../UPLOADS/NW_ST1/\"+samp+\"_R2.fastq.gz\", \"rt\"): \n",
    "    #if iprog > 1000000: \n",
    "    #   break; \n",
    "    rd2ct += 0.25\n",
    "    if iprog % 1000000 == 0: \n",
    "       print ([\"Counting R2\", samp, iprog])\n",
    "    iprog += 0.25\n",
    "\n",
    "\n",
    "statsPrbs[\"TotReads\\tFastq\"]=rd2ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading R2\tNW072-WT1-210123_S1_L001\t0\t0%\n"
     ]
    }
   ],
   "source": [
    "#### Process R2 (Determine if probe was present and if it was digested and completed properly ) \n",
    "iprog=0\n",
    "fl=\"../UPLOADS/NW_ST2/\"+samp+\"_R2.sam.gz\"\n",
    "for [rdid,flag,genome, chromo,xstart, xend,mapq,mapqbin, seq, sam] in SamReader(fl): \n",
    "    #if iprog > 1000000: \n",
    "    #   break; \n",
    "    if iprog % 1000000 == 0: \n",
    "       print( \"\\t\".join(map(str, [\"Loading R2\", samp, iprog, str(int(100*iprog/rd2ct))+\"%\", ]))) \n",
    "    iprog += 1    \n",
    "    [a,b] =  CheckProbe(chromo,xstart,xend, flag)\n",
    "    if b == \"VALID\":\n",
    "       valid[rdid] = a\n",
    "    if a != \"UNKNOWN\": \n",
    "       if a+\"\\t\"+b in statsPrbs: \n",
    "          statsPrbs[a+\"\\t\"+b] += 1\n",
    "       else: \n",
    "          statsPrbs[a+\"\\t\"+b] = 1    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> The program is only getting the sam ID from R2 and perform the valid operation only on the R1 samfile ID, such as rdid='M00559:174:000000000-CPFMJ:1:1101:14331:1816'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### Process R1 (Look for the complementary sequence) \n",
    "#print(statsPrbs)\n",
    "fl=\"../UPLOADS/NW_ST2/\"+samp+\"_R1.sam.gz\"\n",
    "statsFrags={}\n",
    "iprog=1\n",
    "ValidCov={}\n",
    "for [rdid,flag,genome, chromo,xstart, xend,mapq, mapqbin, seq, sam] in SamReader(fl): \n",
    "    #if iprog > 1000000: \n",
    "    #   break; \n",
    "    if iprog % 1000000 == 0: \n",
    "       print( \"\\t\".join(map(str, [\"Loading R1\", samp, iprog, str(int(100*iprog/rd2ct))+\"%\", ]))) \n",
    "    iprog += 1\n",
    "    isValid = rdid in valid\n",
    "    statid=\"\\t\".join([samp,\"R1\",genome, chromo, str(1000000*((xstart + xend)/2000000)), mapqbin, str(isValid)])\n",
    "    if statid in statsLib: \n",
    "       statsLib[statid] += 1\n",
    "    else: \n",
    "       statsLib[statid] = 1\n",
    "    if isValid and mapqbin not in [\"BlackHole\", \"UNKNOWN\",\"UNMAPPED\", \"AMB\", \"LT10\"]:\n",
    "       frg = getREFrag(genome+\":\"+chromo,int((xstart+xend)/2),RE)\n",
    "       prb = valid[rdid]       \n",
    "       vstat = \"\\t\".join( map(str, [prb,flag,genome, chromo,xstart, xend,frg[0]] )  )\n",
    "       if not vstat in ValidCov: \n",
    "          ValidCov[vstat] = 1\n",
    "       else: \n",
    "          ValidCov[vstat] += 1 \n",
    "       entryid = frg[0] + \"\\t\" + prb\n",
    "       if not entryid in statsFrags: \n",
    "          statsFrags[entryid] = {\"LT10\":0, \"LT30\":0, \"GT30\":0, \"RE_Start\":frg[1], \"RE_End\":frg[2],\"cov\":{},\"umi\":{}}      \n",
    "       if useUMI and mapq > 10: \n",
    "          u = umis[rdid]\n",
    "          if u in statsFrags[entryid][\"umi\"]:\n",
    "             statsFrags[entryid][\"umi\"][u] += 1 \n",
    "          else: \n",
    "             statsFrags[entryid][\"umi\"][u] = 1 \n",
    "       statsFrags[entryid][mapqbin] += 1\n",
    "       iPos=\"XX\"\n",
    "       if flag == \"0\":           \n",
    "          iPos = \"F:\"+ str(frg[2] - xstart)\n",
    "       if flag == \"16\": \n",
    "          iPos = \"R:\"+ str(xend - frg[1])\n",
    "       if iPos in statsFrags[entryid][\"cov\"]:\n",
    "          statsFrags[entryid][\"cov\"][iPos] += 1\n",
    "       else: \n",
    "          statsFrags[entryid][\"cov\"][iPos] = 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################################################################################################################\n",
    " <p style=\"color:blue;font-size:30px\"> ##Write Results </p>\n",
    " \n",
    " \n",
    " \n",
    "<ul style=\"color:black;font-size:20px\"> <p> The program will write the following output files </p>\n",
    "<li>../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ProbeStats.tab\n",
    "<li>../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ReadStats.tab\n",
    "<li>../UPLOADS/STATS/NW072-WT1-210123_S1_L001_FragStats.tab\n",
    "<li>../UPLOADS/STATS/NW072-WT1-210123_S1_L001_UmiStats.tab\n",
    "<li>../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ValidCovStats.tab\n",
    "    </ul>\n",
    "############################################################################################################################\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output file : \n",
      "../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ProbeStats.tab\n",
      "../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ReadStats.tab\n",
      "../UPLOADS/STATS/NW072-WT1-210123_S1_L001_FragStats.tab\n",
      "../UPLOADS/STATS/NW072-WT1-210123_S1_L001_UmiStats.tab\n",
      "../UPLOADS/STATS/NW072-WT1-210123_S1_L001_ValidCovStats.tab\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "out = open(\"../UPLOADS/STATS/\"+samp+\"_ProbeStats.tab\",\"w\")\n",
    "for x in statsPrbs: \n",
    "    out.write(samp + \"\\t\" + x + \"\\t\" + str(statsPrbs[x]) + \"\\n\")\n",
    "    \n",
    "print(\"Output file : \")\n",
    "\n",
    "print(\"../UPLOADS/STATS/\"+samp+\"_ProbeStats.tab\")\n",
    "\n",
    "out.close()\n",
    "\n",
    "\n",
    "out = open(\"../UPLOADS/STATS/\"+samp+\"_ReadStats.tab\",\"w\")\n",
    "for x in statsLib: \n",
    "    out.write(x + \"\\t\" + str(statsLib[x]) + \"\\n\")\n",
    "\n",
    "print(\"../UPLOADS/STATS/\"+samp+\"_ReadStats.tab\")\n",
    "out.close()\n",
    "\n",
    "\n",
    "out = open(\"../UPLOADS/STATS/\"+samp+\"_FragStats.tab\",\"w\")\n",
    "outUMI = open(\"../UPLOADS/STATS/\"+samp+\"_UmiStats.tab\",\"w\")\n",
    "out.write( \"\\t\".join([\"SAMP\",\"Frag\",  \"GENOME\",\"CHROMO\", \"FragStart\",\"FragEnd\",\"PROBE\",\"ProbeChromo\",\"ProbePosition\",\"ProbeGenome\" ,\"LT10\",\"LT30\",\"GT30\", \"ctpos\",\"ctPos5\",\"ctPos10\",\"ctumi\",\"ctUmi5\",\"ctUmi10\",\"maxPrbMatches\", \"maxPrbOverPos\", \"maxPrbTp\",\"Positions\",\"Umis\"]) + \"\\n\")\n",
    "print(\"../UPLOADS/STATS/\"+samp+\"_FragStats.tab\")\n",
    "print(\"../UPLOADS/STATS/\"+samp+\"_UmiStats.tab\")\n",
    "\n",
    "\n",
    "for x in statsFrags:\n",
    "    [frag,prb ]=x.split(\"\\t\")\n",
    "    [genome, chromo, RE_frag] = frag.split(\":\")\n",
    "    frgStats = statsFrags[x]\n",
    "    LT10=frgStats[\"LT10\"]\n",
    "    LT30=frgStats[\"LT30\"]\n",
    "    GT30=frgStats[\"GT30\"]\n",
    "    FragStart=frgStats[\"RE_Start\"]\n",
    "    FragEnd=frgStats[\"RE_End\"]    \n",
    "    ctpos=len(frgStats[\"cov\"])\n",
    "    if useUMI: \n",
    "       ctumi=len(frgStats[\"umi\"])\n",
    "    else: \n",
    "       ctumi=0\n",
    "    maxPrbMatches=0 \n",
    "    maxPrbOverPos=0 \n",
    "    maxPrbTp=\"UNK\"\n",
    "    seq = refGenome[genome+\":\"+chromo][FragStart:FragEnd] \n",
    "    maxPrbMatches=0\n",
    "    maxPrbOverPos=-1\n",
    "    maxPrbTp=\"NONE\"\n",
    "    probe = prbsALL[prb]\n",
    "    prb_seq = refGenome[probe[6] + \":\"+probe[1] ][probe[3]:probe[4]]\n",
    "    for i in range(1, len(seq) - len(prb_seq) - 1):\n",
    "           subseq= seq[i:(len(prb_seq)+i)  ]\n",
    "           dst = 1.0*distance(prb_seq, subseq) \n",
    "           score = 1 - dst/len(prb_seq)\n",
    "           if score > maxPrbMatches:\n",
    "              maxPrbMatches = score\n",
    "              maxPrbOverPos = i + FragStart\n",
    "              maxPrbTp=\"FOR\"\n",
    "    if len(seq) > 500: \n",
    "       seq=\"Too Long\"\n",
    "       covct=\"Too Long\"\n",
    "    ctPos5 = 0\n",
    "    ctPos10 = 0\n",
    "    xpos = \"\"\n",
    "    for x in frgStats[\"cov\"]:  \n",
    "      if frgStats[\"cov\"][x] > 5: \n",
    "         xpos += x+\":\"+str(frgStats[\"cov\"][x])+\",\"\n",
    "         ctPos5 += 1\n",
    "      if frgStats[\"cov\"][x] > 10: \n",
    "         ctPos10 += 1\n",
    "    xpos = xpos[0:2000]\n",
    "    xumi = \"\"\n",
    "    ctUmi5 = 0\n",
    "    ctUmi10 = 0\n",
    "    for u in frgStats[\"umi\"]:  \n",
    "      xx=[samp,frag,prb,u,str(frgStats[\"umi\"][u]), str(ctpos) ]\n",
    "      outUMI.write(\"\\t\".join(xx)+\"\\n\") \n",
    "      if frgStats[\"umi\"][u] > 5: \n",
    "         xumi += u+\":\"+str(frgStats[\"umi\"][u])+\",\"\n",
    "         ctUmi5 += 1\n",
    "      if frgStats[\"umi\"][u] > 10: \n",
    "         ctUmi10 += 1\n",
    "    out.write( \"\\t\".join(map(str,[samp, frag,  genome, chromo, FragStart,FragEnd,prb,probe[1],probe[3],probe[6],LT10,LT30,GT30, ctpos,ctPos5,ctPos10, ctumi,ctUmi5,ctUmi10, maxPrbMatches, maxPrbOverPos, maxPrbTp, xpos, xumi])) + \"\\n\")\n",
    "\n",
    "\n",
    "out.close()\n",
    "outUMI.close()\n",
    "\n",
    "out = open(\"../UPLOADS/STATS/\"+samp+\"_ValidCovStats.tab\",\"w\")\n",
    "for x in ValidCov: \n",
    "  ln= x+\"\\t\"+str(ValidCov[x])+\"\\n\"\n",
    "  out.write(ln)\n",
    "\n",
    "print(\"../UPLOADS/STATS/\"+samp+\"_ValidCovStats.tab\")\n",
    "out.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> Distance and parseCigar functions </h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distancex(seq1, seq2):\n",
    "    oneago = None\n",
    "    thisrow = range(1, len(seq2) + 1) + [0]\n",
    "    for x in xrange(len(seq1)):\n",
    "        twoago, oneago, thisrow = oneago, thisrow, [0] * len(seq2) + [x + 1]\n",
    "        for y in xrange(len(seq2)):\n",
    "            delcost = oneago[y] + 1\n",
    "            addcost = thisrow[y - 1] + 1\n",
    "            subcost = oneago[y - 1] + (seq1[x] != seq2[y])\n",
    "            thisrow[y] = min(delcost, addcost, subcost)\n",
    "    return thisrow[len(seq2) - 1]\n",
    "def parseCigar(cigar):\n",
    "    chars=[\"M\",\"N\",\"D\",\"I\",\"S\",\"H\",\"P\",\"X\",\"=\"]    \n",
    "    cts={}\n",
    "    temp=\"\"\n",
    "    x=0\n",
    "    for i in range(0,len(cigar)): \n",
    "        if not cigar[i] in chars: \n",
    "           temp +=  cigar[i]\n",
    "        else:            \n",
    "           cts[len(cts)] = [cigar[i],int(temp)]\n",
    "           if cigar[i] in cts: \n",
    "              cts[cigar[i]] += int(temp)\n",
    "           else: \n",
    "              cts[cigar[i]] = int(temp)  \n",
    "           temp=\"\"\n",
    "    return cts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################################################################################################################\n",
    "##Analysis Functions \n",
    "############################################################################################################################\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def SamReader(fl):\n",
    "    i = 0\n",
    "    for rd in gzip.open(fl,\"rt\"): \n",
    "      if rd[0:3] != \"@SQ\": \n",
    "        #M00559:63:000000000-A5KK7:1:1101:10703:1212     16      HG19:chr16      49386963        179     106M45S *       0       0       SEQ QUAL AS:i:102        XS:i:0  XF:i:3  XE:i:2  NM:i:2\n",
    "        cols = rd[:-1].split(\"\\t\")\n",
    "        [segid,flag,chromo,pos,mapq,cigar] = cols[0:6]\n",
    "        seq=cols[9]\n",
    "        mapqbin=\"UNKNOWN\"\n",
    "        mapq = int(mapq)\n",
    "        if flag == \"4\": \n",
    "           mapqbin = \"UNMAPPED\"\n",
    "        elif mapq == 0: \n",
    "           mapqbin=\"AMB\"\n",
    "        elif mapq < 10:\n",
    "           mapqbin = \"LT10\"\n",
    "        elif mapq < 30:     \n",
    "           mapqbin = \"LT30\"\n",
    "        else:  \n",
    "           mapqbin = \"GT30\"                \n",
    "        if mapq > 0: \n",
    "           matched = parseCigar(cigar)['M']\n",
    "           [genome,chromo] = chromo.split(\":\")\n",
    "           if chromo+\":\"+str(int(pos)/1000) in lstBlackListed: \n",
    "              mapqbin = \"BlackHole\"\n",
    "        else: \n",
    "           matched = 0\n",
    "           [genome,chromo] = [\"NA\",\"NA\"]\n",
    "        i += 1      \n",
    "        yield  [segid,flag,genome,chromo,int(pos), int(pos) + matched,mapq,mapqbin, seq, rd] \n",
    "  \n",
    "\n",
    "def SamReader2(fl):\n",
    "    i = 0\n",
    "    #print(fl)\n",
    "    samfile = pysam.AlignmentFile(fl, \"rb\")\n",
    "    for read in samfile.fetch():\n",
    "      rd = read.to_string()  \n",
    "      if rd[0:3] != \"@SQ\": \n",
    "        #M00559:63:000000000-A5KK7:1:1101:10703:1212     16      HG19:chr16      49386963        179     106M45S *       0       0       SEQ QUAL AS:i:102        XS:i:0  XF:i:3  XE:i:2  NM:i:2\n",
    "        cols = rd[:-1].split(\"\\t\")\n",
    "        [segid,flag,chromo,pos,mapq,cigar] = cols[0:6]\n",
    "        seq=cols[9]\n",
    "        mapqbin=\"UNKNOWN\"\n",
    "        mapq = int(mapq)\n",
    "        if flag == \"4\": \n",
    "           mapqbin = \"UNMAPPED\"\n",
    "        elif mapq == 0: \n",
    "           mapqbin=\"AMB\"\n",
    "        elif mapq < 10:\n",
    "           mapqbin = \"LT10\"\n",
    "        elif mapq < 30:     \n",
    "           mapqbin = \"LT30\"\n",
    "        else:  \n",
    "           mapqbin = \"GT30\"                \n",
    "        if mapq > 0: \n",
    "           matched = parseCigar(cigar)['M']\n",
    "           [genome,chromo] = chromo.split(\":\")\n",
    "           if chromo+\":\"+str(int(pos)/1000) in lstBlackListed: \n",
    "              mapqbin = \"BlackHole\"\n",
    "        else: \n",
    "           matched = 0\n",
    "           [genome,chromo] = [\"NA\",\"NA\"]\n",
    "        i += 1      \n",
    "        yield  [segid,flag,genome,chromo,int(pos), int(pos) + matched,mapq,mapqbin, seq, rd] \n",
    "\n",
    "\n",
    "def CheckProbe(chromo,xstart,xend, flag):\n",
    "    pos1kbp = chromo+\":\"+str(int(xstart/1000))\n",
    "    if pos1kbp not in probes:\n",
    "       return [\"UNKNOWN\",\"UNKNOWN\"]\n",
    "    bestDst = 10000\n",
    "    #Find the probe closest to the read start\n",
    "    bestEntry = []\n",
    "    for [prbid,prbStart,prbEnd,prbFragEnd,prbTp] in probes[pos1kbp]: \n",
    "        if prbTp == \"+\": \n",
    "           xdst = abs(xstart - prbStart) \n",
    "        else:\n",
    "           xdst = abs(xend - prbStart) \n",
    "        if xdst < bestDst: \n",
    "           bestDst = xdst\n",
    "           bestEntry = [prbid,prbStart,prbEnd,prbFragEnd,prbTp]\n",
    "          # print('the best entry is',bestEntry)\n",
    "    [prbid,prbStart,prbEnd,prbFragEnd,prbTp] = bestEntry\n",
    "    if prbTp == \"+\": \n",
    "       if flag != \"0\":\n",
    "          return [prbid,\"FLAG-Mismatch:\"+flag]\n",
    "       else:  \n",
    "          if abs(xstart - prbStart) < 5 and abs(xend - prbFragEnd) < 5: \n",
    "             return [prbid,\"VALID\"]\n",
    "          elif abs(xstart - prbStart) < 5 and abs(xend - prbEnd) < 5:  \n",
    "             return [prbid,\"MISSANEAL\"] \n",
    "          elif abs(xstart - prbStart) < 5 and xend > prbFragEnd:  \n",
    "             return [prbid,\"UNDIG\"]\n",
    "          else: \n",
    "             return [prbid,\"NO-MATCH\"]\n",
    "    else: \n",
    "       if flag != \"16\":\n",
    "          return [prbid,\"FLAG-Mismatch:\"+flag]\n",
    "       else: \n",
    "          if abs(xend - prbStart) < 5 and abs(xstart - prbFragEnd) < 5: \n",
    "             return [prbid,\"VALID\"]\n",
    "          elif abs(xend - prbStart) < 5 and abs(xstart - prbEnd) < 5:  \n",
    "             return [prbid,\"MISSANEAL\"] \n",
    "          elif abs(xend - prbStart) < 5 and xstart < prbFragEnd:  \n",
    "             return [prbid,\"UNDIG\"]\n",
    "          else: \n",
    "             return [prbid,\"NO-MATCH\"]    \n",
    "\n",
    "\n",
    "def getREFrag(chromo,  pos, resEnz): \n",
    "   B = refGenome[chromo].find(resEnz, pos -  len(resEnz) + 1)\n",
    "   if B > 0: \n",
    "      B = B + len(resEnz) \n",
    "   else: \n",
    "      B = len(refGenome[chromo]) \n",
    "   A = -1\n",
    "   offset = 0\n",
    "   while A < 0 and pos > offset: \n",
    "      A = refGenome[chromo].rfind(resEnz, pos - offset - 1010, pos - offset )\n",
    "      offset += 1000\n",
    "   A=min(max(1,A), len(refGenome[chromo]))\n",
    "   B=min(max(1,B), len(refGenome[chromo]))\n",
    "   return [chromo + \":\" + str(A) + \"+\" + str(B-A), A,B]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
